Provide new default for runtime identifier Fixes #35575

Forgind opened the pull request:
Recreated #36149 to avoid pinging everyone as I changed the base from 8.0.2xx to main.

Fixes https://github.com/dotnet/sdk/issues/35575
 
This block of code essentially set a default for the runtime identifier if none was specified. It did not take into account that, when users do not specifically request legacy rids, we shifted the default to use the simplified rid graph. This now does that.

Forgind committed:
Also exempt RuntimeIdentifiers

Forgind committed:
Conditionally switch default to win-x*

Forgind committed:
Use homegrown property

Forgind committed:
Merge branch 'main' of https://github.com/dotnet/sdk into augment-condition

Forgind committed:
Disable on FullFx

dsplaisted commented:
It looks like the PR description doesn't entirely match the changes in the PR.  Could you update that?

It looks like this is introducing a new `UseLegacyRids` property.  Instead of that, could we key off the existing `UseRidGraph` property?

Could you add test cases for this?

Forgind committed:
Switch to using UseRidGraph

Forgind committed:
Increase version

Forgind committed:
Swap in Libuv for SQLite

SQLite doesn't have a later version on nuget.org, and 3.13.0 doesn't support win-x*

Forgind committed:
Make test

Forgind committed:
Delete space

Forgind committed:
Merge branch 'main' of https://github.com/dotnet/sdk into augment-condition-2

Forgind committed:
Create and load the project directly

Forgind committed:
Try a slow search :(

joeloff commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:
```diff
@@ -54,11 +54,13 @@ Copyright (c) .NET Foundation. All rights reserved.
     <_UsingDefaultRuntimeIdentifier>true</_UsingDefaultRuntimeIdentifier>
     <RuntimeIdentifier Condition="'$(PlatformTarget)' == 'x64'">win7-x64</RuntimeIdentifier>
     <RuntimeIdentifier Condition="'$(PlatformTarget)' == 'x86' or '$(PlatformTarget)' == ''">win7-x86</RuntimeIdentifier>
+    <RuntimeIdentifier Condition="'$(PlatformTarget)' == 'x64' and '$(UseRidGraph)' != 'true'">win-x64</RuntimeIdentifier>
```
Should we check for `win-arm64`?

dsplaisted commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs:
```diff
@@ -33,12 +36,83 @@ public void It_builds_a_simple_desktop_app()
+                new global::Microsoft.Build.Logging.ConsoleLogger(LoggerVerbosity.Detailed)
+            };
+
+                var collection = new ProjectCollection(null, loggers, ToolsetDefinitionLocations.Default);
```
We shouldn't use MSBuild APIs like this to load a test project.  That mixes up stage 0 and stage 2.  Rather than redirect the in-process stage 0 MSBuild to use some pieces of stage 2, we should use stage 2 in a separate process to do the evaluation.  You can use the `GetValuesCommand` class for this.

dsplaisted reviewed (approved):
Code changes look good.  See my comments for test changes that I recommend.

Forgind commented:
(I talked about these changes with dsplaisted, and he said it was ok to merge as-is with possible future work to change how we write the test.)
